opam compiles the wrong sources after overwriting a pin

opam somehow managed to fetch the wrong sources for one of the packages. I wanted to downgrade `coq` from `8.9.dev` to `8.8.2`, and what opam ended up doing is compile the `8.9.dev` sources with the opam files for `8.8.2`. That didn't work out so well because some binaries got removed since then.

Basically, https://github.com/ocaml/opam/issues/3262 is back.

Here's the commands I ran:
```
opam install coq-stdpp.1.1.0 # failed because not compatible with coq.8.9.dev
opam install coq.8.8.2 # failed because 8.9.dev is pinned
opam pin coq 8.8.2 # replaced the pin but failed to build because it used the wrong sources
```

The output:
```
r@r-thinktop:~ [coq-test]$ opam install coq.8.8.2

<><> Synchronising pinned packages ><><><><><><><><><><><><><><><><><><><><><><>
[coq.8.9.dev] synchronised from git+https://github.com/coq/coq.git#v8.9

[ERROR] coq = 8.8.2 not available because the package is pinned to version 8.9.dev
r@r-thinktop:~ [coq-test]$ opam pin coq 8.8.2
[NOTE] Package coq used to be pinned to version 8.9.dev
coq is now pinned to version 8.8.2

The following actions will be performed:
  ↘ downgrade coq               8.9.dev to 8.8.2*
  ↻ recompile coq-stdpp         dev.2018-10-15.1.b7f1f6b1 [uses coq]
  ↻ recompile coq-iris-builddep dev*                      [uses coq]
===== ↻ 2   ↘ 1 =====
Do you want to continue? [Y/n] y

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[coq-stdpp.dev.2018-10-15.1.b7f1f6b1] no changes from git+https://gitlab.mpi-sws.org/robbertkrebbers/coq-stdpp.git#b7f1f6b178091ca6e3d363208d7875d73321903c

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
⊘ removed   coq-iris-builddep.dev
⊘ removed   coq-stdpp.dev.2018-10-15.1.b7f1f6b1
⊘ removed   coq.8.9.dev
[ERROR] Installation of coq.8.8.2 failed

#=== ERROR while installing coq.8.8.2 =========================================#
Some files in /home/r/.opam/coq-test/.opam-switch/install/coq.install couldn't be installed:
  - /home/r/.opam/coq-test/.opam-switch/build/coq.8.8.2/bin/gallina to /home/r/.opam/coq-test/bin



<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
┌─ The following actions failed
│ ∗ install coq 8.8.2
└─ 
┌─ The following changes have been performed (the rest was aborted)
│ ⊘ remove coq               8.9.dev
│ ⊘ remove coq-iris-builddep dev
│ ⊘ remove coq-stdpp         dev.2018-10-15.1.b7f1f6b1
└─ 
[NOTE] Pinning command successful, but your installed packages may be out of sync.

The former state can be restored with:
    opam switch import "/home/r/.opam/coq-test/.opam-switch/backup/state-20181103173624.export"
Or you can retry to install your package selection with:
    opam install --restore
```

I guess I could fix this switch by deleting `/home/r/.opam/coq-test/.opam-switch/build/coq.8.8.2` but I do not need this switch so I will just leave it as-is.

```
# opam config report
# opam-version      2.0.1 
# self-upgrade      no
# system            arch=x86_64 os=linux os-distribution=debian os-version=testing
# solver            builtin-mccs+glpk
# install-criteria  -removed,-count[version-lag,request],-count[version-lag,changed],-changed
# upgrade-criteria  -removed,-count[version-lag,solution],-new
# jobs              8
# repositories      4 (http), 4 (local), 1 (version-controlled) (default repo at 7ed1cc7a)
# pinned            1 (git), 1 (rsync), 1 (version)
# current-switch    coq-test

```

--- a/src/client/opamAction.ml
+++ b/src/client/opamAction.ml
@@ -148,8 +148,14 @@ let download_package st nv =
   else
   let dir = OpamSwitchState.source_dir st nv in
   if OpamPackage.Set.mem nv st.pinned &&
-     OpamFilename.exists_dir dir
-  then Done None
+     OpamFilename.exists_dir dir &&
+     OpamStd.Option.Op.(
+       OpamPinned.find_opam_file_in_source nv.name dir >>=
+       OpamFile.OPAM.read_opt >>=
+       OpamFile.OPAM.version_opt)
+     = Some nv.version
+  then
+    Done None
   else
     (OpamUpdate.cleanup_source st
        (OpamPackage.Map.find_opt nv st.installed_opams)

